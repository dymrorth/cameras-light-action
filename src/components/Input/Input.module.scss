@import 'styles/common';

.inputWrap {
    position: relative;

    .input {
        position: relative;
        background: none;
        border: none;
        border-bottom: 4px solid var(--color-text-1);
        color: var(--color-text-1);
        font-size: $font-size-xl;
        transition: border var(--transition-duration) ease;
        outline: none;
    }

    .icon {
        position: absolute;
        top: 0;
        right: 0;

        svg {
            font-size: $font-size-xl;
        }
    }

    &:hover {
        .input {
            border-color: var(--color-primary-dark);
        }
        .icon {
            color: var(--color-primary-dark);
        }
    }

    &:active,
    &:focus-within {
        .input {
            border-color: var(--color-primary);
        }
        .icon {
            color: var(--color-primary);
        }
    }
}
