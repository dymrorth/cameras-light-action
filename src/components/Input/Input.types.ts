export interface InputProps {
    className?: string
    icon?: React.ReactNode
    onChange?: React.ChangeEventHandler<HTMLInputElement>
    placeholder?: string
}
