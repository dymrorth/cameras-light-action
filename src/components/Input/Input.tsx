import React from 'react'

import { InputProps } from './Input.types'
import styles from './Input.module.scss'

import classNames from 'classnames'

const Input: React.FC<InputProps> = ({ className, icon, ...props }) => {
    const inputWrapClasses = classNames(styles.inputWrap, className)

    return (
        <div className={inputWrapClasses}>
            <input type="text" {...props} className={styles.input} />
            {icon && (
                <div className={styles.icon}>
                    {icon}
                </div>
            )}
        </div>
    )
}

export default Input
