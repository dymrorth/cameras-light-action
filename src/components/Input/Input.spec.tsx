import Input from './Input'

import { render, RenderResult, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

describe('<Input />', () => {
    describe('when mounted', () => {
        it('has placeholder text', () => {
            const placeholderText = 'Search...'
            const { asFragment }: RenderResult = render(<Input placeholder={placeholderText} />)
            const input: HTMLElement = screen.getByPlaceholderText(placeholderText)

            expect(input).toBeDefined()
            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe('with icon', () => {
        it('has icon element', () => {
            const { asFragment }: RenderResult = render(<Input icon={<p>icon</p>} />)
            const input: HTMLElement = screen.getByText(/icon/i)

            expect(input).toBeDefined()
            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe('when typing in', () => {
        it('calls handler function', () => {
            const handleClick = jest.fn()
            const placeholderText = 'Search...'
            const inputTyping = 'this is a placeholder text'

            render(<Input placeholder={placeholderText} onChange={handleClick} />)
            const input: HTMLElement = screen.getByPlaceholderText(placeholderText)

            userEvent.type(input, inputTyping)
            expect(handleClick).toHaveBeenCalledTimes(inputTyping.length)
            expect(input).toHaveValue(inputTyping)
        })
    })
})
