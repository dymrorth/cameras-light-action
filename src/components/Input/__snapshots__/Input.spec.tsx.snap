// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> when mounted has placeholder text 1`] = `
<DocumentFragment>
  <div
    class="inputWrap"
  >
    <input
      class="input"
      placeholder="Search..."
      type="text"
    />
  </div>
</DocumentFragment>
`;

exports[`<Input /> with icon has icon element 1`] = `
<DocumentFragment>
  <div
    class="inputWrap"
  >
    <input
      class="input"
      type="text"
    />
    <div
      class="icon"
    >
      <p>
        icon
      </p>
    </div>
  </div>
</DocumentFragment>
`;
